Speedreader refactor readability backend

keur opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/14787

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

gpestana commented on components/speedreader/rust/lib/src/readability/src/dom.rs:
```diff
@@ -134,54 +136,61 @@ pub fn extract_text(handle: &Handle, text: &mut String, deep: bool) {
-        if let Element { ref name, .. } = child.data {
-            let t = name.local.as_ref();
+#[inline]
+pub fn find_nodes_with_tag(handle: &Handle, tags: &[&str], nodes: &mut Vec<Handle>) {
```
Any special reason for having this signature

```rust
pub fn find_nodes_with_tag(handle: &Handle, tags: &[&str], nodes: &mut Vec<Handle>) {}
```

instead of ?

```rust
pub fn find_nodes_with_tag(handle: &Handle, tags: &[&str]) -> Vec<Handle> {}

```

nit, but IMO returning a new object feels more readable and the API is easier to use.

gpestana commented on components/speedreader/rust/lib/src/readability/src/dom.rs:
```diff
@@ -190,62 +199,68 @@ pub fn has_nodes(handle: &Handle, tag_names: &[&'static LocalName]) -> bool {
 }
 
+#[inline]
+pub fn has_single_tag_inside_element(handle: &Handle, tag: &LocalName) -> Option<Handle> {
```
nit: I'd expect a `has_...()` function to return a boolean. If I understood the function correctly, maybe renaming it to `get_single_element_by_tag()` would be more explicit.

gpestana commented on components/speedreader/rust/lib/src/readability/src/extractor.rs:
```diff
@@ -1,133 +1,99 @@
 use url::Url;
+use util::StringUtils;
+
+const NUM_TOP_CANDIDATES: usize = 5;
```
Why is `NUM_TOP_CANDIDATES ` == 5? An inline comment explaining the reason would be helpfult.

gpestana commented on components/speedreader/rust/lib/src/readability/src/extractor.rs:
```diff
@@ -1,133 +1,99 @@
-    }
-
-    let mut id: &str = "/";
+    scorer::find_candidates(&mut dom, handle.clone());
```
Any specific reason for this clone?. I'd pass the `handle` object only.

gpestana reviewed (changes requested):
`cargo test` does not work because the integration tests (`lib/tests/*`) are still using the old implementation.

I'm still going to do another pass on the PR, but added a few comments that can be addressed for now.

keur commented on components/speedreader/rust/lib/src/readability/src/dom.rs:
```diff
@@ -1,56 +1,57 @@
     }
 }
 
+#[inline]
```
Doubt it makes a difference at all. I just made the smaller non-recursive functions that get called a lot inline.

keur replied:
I'm all for benchmarking, but let's also make sure the algorithm is fully complete, we get great results, then we can do an optimization pass.

keur commented:
@gpestana All suggestions made, and fixed up the tests using the markup5ever dom, so they pass now.

keur commented:
Skipping CI because the rust tests haven't been hooked up to GN.

pes10k commented on components/speedreader/rust/lib/examples/feature_extraction_streamer.rs:
```diff
@@ -19,6 +19,11 @@ fn main() {
+    sink.rcdom.on_parse_error = Some(Box::new(|errors| {
+        // parsing errors may happen due to malformed HTML, but it will not affect
+        // parsing itself or the DOM tree building
+        println!("Errors: {:?}", errors);
```
Probably we don't want general print statements in Brave Core.  Is there a LOGV or equiv for rust?

keur replied:
this is inside the example module. core doesn't call this code. just keeping parity with the old code since i ported the DOM to a newer (safer) library.

keur replied:
The important files for the Arc90 stuff are `src/readability/extractor.rs`, `src/readability/scorer.rs`, and `src/readability/dom.rs` (utils for the scorer).

pes10k reviewed (approved):
I'm marking approve, but please take with a grain of salt, since my rust is very very baby level

pes10k commented on components/speedreader/rust/lib/Cargo.toml:
```diff
@@ -18,15 +18,15 @@ include = [
 html5ever = "0.25.1"
-markup5ever_rcdom = "0.1.0"
-regex = "1"
+kuchiki = { git = "https://github.com/keur/kuchiki", branch = "speedreader" }
```
If the repo is going to be included in brave builds, should the repo be moved or forked in the brave GH org, and possibly reviewed by sec team?

keur replied:
Good point. Wasn't sure what to do here. So here is the difference between my branch and the master branch (very minor) can be seen here: https://github.com/kuchiki-rs/kuchiki/compare/master...keur:speedreader. It exposes some fields that were public in `Rcdom` and adds two metadata fields to massively simplify this implementation. No logic changes to the algorithm or anything like that. 

The existing implementation relied on the DOM being introspectable by API callers. Especially in the classification pass. Doesn't seem like there's any way around it, unfortunately. I tried to send these up, but upstream said if we are going to be doing this much DOM customization, we might as well have our own fork. https://github.com/kuchiki-rs/kuchiki/pull/87#issuecomment-803269129. So I went forward with that. I plan to keep my changes on top of their DOM extremely minimal.

keur replied:
Moving this to brave org is a good idea. Someone with perms just needs to fork https://github.com/kuchiki-rs/kuchiki/ and give me write access to that.

pes10k replied:
Sounds good. I think asking in #security-discussion is a great idea, since they can fork it, and review it before hand if they think its necessary

keur replied:
Moved to Brave org in latest commit.

keur commented on components/speedreader/rust/lib/src/readability/tests/lib.rs:
```diff
@@ -64,7 +64,7 @@ mod test {
 //test!(engadget);
 //test!(folha);
 //test!(gmw);
-test!(guardian_1);
```
this test has been failing since before my change. these tests aren't very good. they just have output and expected htmls, which will break at any change in the algorithm. i moved unit tests into the extractor.rs file, and tested specific features of the algorithm.

gpestana commented on components/speedreader/rust/lib/src/readability/src/util.rs:
```diff
@@ -0,0 +1,46 @@
+    elem_index(v, Ordering::Greater)
+}
+
+pub trait StringUtils {
```
Any special reason to define this as a trait?

gpestana reviewed (approved):
Some nits here and there and a couple of questions, but in general the code looks good. All things considered, I believe we should keep the E2E tests (even if we cut some of the test cases).

Also it would be great to run `cargo fmt`

keur commented on components/speedreader/rust/lib/src/readability/src/dom.rs:
```diff
@@ -190,62 +201,68 @@ pub fn has_nodes(handle: &Handle, tag_names: &[&'static LocalName]) -> bool {
+    let mut elems = handle
+        .children()
+        .filter(|child| child.as_element().is_some());
+    if elems.clone().count() == 1 {
```
Agreed. So the point of this part is sites like mobile.slate.com wrap every paragraph element in a div, and this can confuse the scoring algorithm. So we hoist those paragraph elements up and remove the divs. So I think the most clear thing we can do is rename the function as you suggested, and move that elemet count == 1 check into the scorer.rs, since it has more to do with the algorithm.

keur commented on components/speedreader/rust/lib/src/readability/src/dom.rs:
```diff
@@ -1,56 +1,57 @@
-        Element { ref attrs, .. } => attr(name, &attrs.borrow()),
-        _ => None,
+#[inline]
+pub fn clone_element_steal_children(
```
going to comment all these functions as discussed in DM.

keur commented:
@gpestana made all your changes, added comments to every function in `scorer.rs` and `dom.rs`. If you grab the test data (discussed in DM), there are still many integration tests that will run. I agree we should have E2E tests in the readability subtree too, but will fix that in future PR since this one is getting to be huge. The only test that I removed has been failing since before I made any changes, so we'll refactor everything in there.
